File: CoProcessing/Adaptors/PhastaAdaptor/phastaadaptor.f
c$$$=========================================================================
c$$$
c$$$  Program:   ParaView
c$$$  Module:    phastaadaptor.f
c$$$
c$$$  Copyright (c) Kitware, Inc.
c$$$  All rights reserved.
c$$$  See Copyright.txt or http://www.paraview.org/HTML/Copyright.html for details.
c$$$
c$$$     This software is distributed WITHOUT ANY WARRANTY; without even
c$$$     the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
c$$$     PURPOSE.  See the above copyright notice for more information.
c$$$
c$$$=========================================================================


c...==============================================================
c... subroutine to do the coprocessing
c... The subroutine is responsible for determining if coprocessing
c... is needed this timestep and if it is needed then the
c... subroutine passes the phasta data structures into
c... the coprocessor. This is meant to be called at the end of
c... every time step.
c... The input is:
c...    itimestep -- the current simulation time step
c...    X -- the coordinates array of the nodes
c...    Y -- the fields array (e.g. velocity, pressure, etc.)
c...    compressibleflow -- flag to indicate whether or not the
c...                         flow is compressible.  if it is then
c...                         temperature will be outputted
c... It has no output and should not change any Phasta data.
c...==============================================================

      subroutine phastacoprocessor(itimestep, X, Y, compressibleflow)
      use pointer_data
      include "common.h"
      integer iblk, nenl, npro, j, needflag
      integer compressibleflow, itimestep
      dimension x(numnp,nsd), y(nshg,ndof)

      if(nshg .ne. numnp) then
         print *, 'CoProcessing only setup for when nshg equals numnp'
         return
      endif
c  First check if we even need to coprocess this time step/time
      call requestdatadescription(itimestep, time, needflag)
      if(needflag .eq. 0) then
c  We don't need to do any coprocessing now so we can return
         return
      endif

c  Check if we need to create the grid for the coprocessor
      call needtocreategrid(needflag)
      if(needflag .ne. 0) then
c     We do need the grid.
         call createpointsandallocatecells(numnp, X, numel)

        do iblk=1,nelblk
            nenl = lcblk(5,iblk) ! no. of vertices per element
            npro = lcblk(1,iblk+1) - lcblk(1,iblk) ! no. of elements in block
            call insertblockofcells(npro, nenl, mien(iblk)%p(1,1))
        enddo
      endif ! if needflag .ne. 0 --

c  Inside addfields we check to see if we really need the field or not
      call addfields(nshg, ndof, Y, compressibleflow)

      call coprocess()

      return
      end

File: Adaptors/Phasta/PhastaAdaptor.cxx
/*=========================================================================

  Program:   ParaView
  Module:    PhastaAdaptor.cxx

  Copyright (c) Kitware, Inc.
  All rights reserved.
  See Copyright.txt or http://www.paraview.org/HTML/Copyright.html for details.

     This software is distributed WITHOUT ANY WARRANTY; without even
     the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
     PURPOSE.  See the above copyright notice for more information.

=========================================================================*/

#include "PhastaAdaptor.h"

#include "FortranAdaptorAPI.h"
#include "PhastaAdaptorAPIMangling.h"

#include "vtkCPAdaptorAPI.h"
#include "vtkCPDataDescription.h"
#include "vtkCPInputDataDescription.h"
#include "vtkCPProcessor.h"
#include "vtkCellData.h"
#include "vtkCellType.h"
#include "vtkDoubleArray.h"
#include "vtkFieldData.h"
#include "vtkPointData.h"
#include "vtkPoints.h"
#include "vtkUnstructuredGrid.h"

//------------------------------------------------------------------------------
void createpointsandallocatecells(int* numPoints, double* coordsArray, int* numCells)
{
  if (!vtkCPAdaptorAPI::GetCoProcessorData())
  {
    vtkGenericWarningMacro("Unable to access CoProcessorData.");
    return;
  }

  vtkUnstructuredGrid* Grid = vtkUnstructuredGrid::New();
  vtkPoints* nodePoints = vtkPoints::New();
  vtkDoubleArray* coords = vtkDoubleArray::New();
  coords->SetNumberOfComponents(3);
  coords->SetNumberOfTuples(*numPoints);
  for (int i = 0; i < *numPoints; i++)
  {
    double tuple[3] = { coordsArray[i], coordsArray[i + *numPoints],
      coordsArray[i + *numPoints * 2] };
    coords->SetTypedTuple(i, tuple);
  }
  nodePoints->SetData(coords);
  coords->Delete();
  Grid->SetPoints(nodePoints);
  nodePoints->Delete();
  Grid->Allocate(*numCells);
  vtkCPAdaptorAPI::GetCoProcessorData()->GetInputDescriptionByName("input")->SetGrid(Grid);
  Grid->Delete();
}

//------------------------------------------------------------------------------
void insertblockofcells(int* numCellsInBlock, int* numPointsPerCell, int* cellConnectivity)
{
  vtkUnstructuredGrid* grid = vtkUnstructuredGrid::SafeDownCast(
    vtkCPAdaptorAPI::GetCoProcessorData()->GetInputDescriptionByName("input")->GetGrid());
  if (!grid)
  {
    vtkGenericWarningMacro("CoProcessing: Could not access grid for cell insertion.");
  }
  int type = -1;
  switch (*numPointsPerCell)
  {
    case 4:
    {
      type = VTK_TETRA;
      break;
    }
    case 5:
    {
      type = VTK_PYRAMID;
      break;
    }
    case 6:
    {
      type = VTK_WEDGE;
      break;
    }
    case 8:
    {
      type = VTK_HEXAHEDRON;
      break;
    }
    default:
    {
      vtkGenericWarningMacro(
        "CoProcessing: Incorrect amount of vertices per element: " << *numPointsPerCell);
      return;
    }
  }
  vtkIdType pts[8]; // assume for now we only have linear elements
  vtkIdType numPoints = grid->GetNumberOfPoints();
  for (int iCell = 0; iCell < *numCellsInBlock; iCell++)
  {
    for (int i = 0; i < *numPointsPerCell; i++)
    {
      pts[i] = cellConnectivity[iCell + i * (*numCellsInBlock)] - 1; //-1 to get from f to c++

      if (pts[i] < 0 || pts[i] >= numPoints)
      {
        vtkGenericWarningMacro(<< pts[i] << " is not a valid node id.");
      }
    }
    if (type == VTK_TETRA)
    { // change the canonical ordering of the tet to match VTK style
      vtkIdType temp = pts[0];
      pts[0] = pts[1];
      pts[1] = temp;
    }
    grid->InsertNextCell(type, *numPointsPerCell, pts);
  }
}

//------------------------------------------------------------------------------
void addfields(int* nshg, int* vtkNotUsed(ndof), double* dofArray, int* compressibleFlow)
{
  vtkCPInputDataDescription* idd =
    vtkCPAdaptorAPI::GetCoProcessorData()->GetInputDescriptionByName("input");
  vtkUnstructuredGrid* UnstructuredGrid = vtkUnstructuredGrid::SafeDownCast(idd->GetGrid());
  if (!UnstructuredGrid)
  {
    vtkGenericWarningMacro("No unstructured grid to attach field data to.");
    return;
  }
  vtkIdType NumberOfNodes = UnstructuredGrid->GetNumberOfPoints();
  // now add numerical field data
  // velocity
  if (idd->IsFieldNeeded("velocity", VTK_DOUBLE))
  {
    vtkDoubleArray* velocity = vtkDoubleArray::New();
    velocity->SetName("velocity");
    velocity->SetNumberOfComponents(3);
    velocity->SetNumberOfTuples(NumberOfNodes);
    for (vtkIdType idx = 0; idx < NumberOfNodes; idx++)
    {
      velocity->SetTuple3(idx, dofArray[idx], dofArray[idx + *nshg], dofArray[idx + *nshg * 2]);
    }
    UnstructuredGrid->GetPointData()->AddArray(velocity);
    velocity->Delete();
  }

  // pressure
  if (idd->IsFieldNeeded("pressure", VTK_DOUBLE))
  {
    vtkDoubleArray* pressure = vtkDoubleArray::New();
    pressure->SetName("pressure");
    pressure->SetArray(dofArray + *nshg * 3, NumberOfNodes, 1);
    UnstructuredGrid->GetPointData()->AddArray(pressure);
    pressure->Delete();
  }

  // Temperature
  // temperature only varies from compressible flow
  if (idd->IsFieldNeeded("temperature", VTK_DOUBLE) && *compressibleFlow == 1)
  {
    vtkDoubleArray* temperature = vtkDoubleArray::New();
    temperature->SetName("temperature");
    temperature->SetArray(dofArray + *nshg * 4, NumberOfNodes, 1);
    UnstructuredGrid->GetPointData()->AddArray(temperature);
    temperature->Delete();
  }
}
